askama_derive/src/generator.rs

```diff
@@ -1295,9 +1302,59 @@ impl<'a> Generator<'a> {
             Expr::RustMacro(name, args) => self.visit_rust_macro(buf, name, args),
             Expr::Try(ref expr) => self.visit_try(buf, expr.as_ref())?,
             Expr::Tuple(ref exprs) => self.visit_tuple(buf, exprs)?,
+            #[cfg(feature = "localization")]
+            Expr::Localize(message, attribute, ref args) => {
+                self.visit_localize(buf, message, attribute, args)?
+            }
         })
     }
 
+    #[cfg(feature = "localization")]
+    fn visit_localize(
+        &mut self,
+        buf: &mut Buffer,
+        message: &str,
+        attribute: Option<&str>,
+        args: &[(&str, Expr<'_>)],
+    ) -> Result<DisplayWrap, CompileError> {
+        let localizer = self.input.localizer.as_ref().expect(
+            "A template struct must have a member with the `#[locale]` \
+             attribute that implements `askama::Localize` to enable calling the localize() filter",
+        );
+
+        let mut message = message.to_string();
+        if let Some(attribute) = attribute {
+            message.push_str(".");
+            message.push_str(attribute);
+        }
+
+        assert!(
+            message.chars().find(|c| *c == '"').is_none(),
+            "message ids with quotes in them break the generator, please remove"
+        );
+
+        self.localized_messages.insert(message.clone());
+
+        buf.write(&format!(
+            "self.{}.translate(\"{}\", &std::iter::FromIterator::from_iter(vec![",
+            localizer.0, message
+        ));
+
+        for (i, (name, value)) in args.iter().enumerate() {
+            if i > 0 {
+                buf.write(", ");
+            }
+            buf.write(&format!(
+                "(\"{}\".to_string(), ({}).into())",
```

Comment:
If you change `Locale::translate()`'s to accept a `HashMap<&str, …>` then you could omit the `.to_string()` call.